server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -1296,27 +1312,45 @@ public static String findV2Template(Metadata metadata, String indexName, boolean
-     * an empty list is returned.
+     * an empty list is returned. If <code>exitOnFirstMatch</code> is true, we return immediately after finding a match.
      */
-    static List<Tuple<String, ComposableIndexTemplate>> findV2CandidateTemplates(Metadata metadata, String indexName, boolean isHidden) {
+    static List<Tuple<String, ComposableIndexTemplate>> findV2CandidateTemplates(
+        Collection<Map.Entry<String, ComposableIndexTemplate>> templateEntries,
+        String indexName,
+        boolean isHidden,
+        boolean exitOnFirstMatch
+    ) {
         final String resolvedIndexName = IndexNameExpressionResolver.DateMathExpressionResolver.resolveExpression(indexName);
-        final Predicate<String> patternMatchPredicate = pattern -> Regex.simpleMatch(pattern, resolvedIndexName);
         final List<Tuple<String, ComposableIndexTemplate>> candidates = new ArrayList<>();
-        for (Map.Entry<String, ComposableIndexTemplate> entry : metadata.templatesV2().entrySet()) {
+        outerLoop:
+        for (Map.Entry<String, ComposableIndexTemplate> entry : templateEntries) {
             final String name = entry.getKey();
             final ComposableIndexTemplate template = entry.getValue();
             if (isHidden == false) {
-                final boolean matched = template.indexPatterns().stream().anyMatch(patternMatchPredicate);
-                if (matched) {
-                    candidates.add(Tuple.tuple(name, template));
+                for (String indexPattern : template.indexPatterns()) {
+                    if (Regex.simpleMatch(indexPattern, resolvedIndexName)) {
+                        candidates.add(Tuple.tuple(name, template));
+                        if (exitOnFirstMatch) {
+                            return candidates;
+                        }
+                    }
                 }
             } else {
-                final boolean isNotMatchAllTemplate = template.indexPatterns().stream().noneMatch(Regex::isMatchAllPattern);
-                if (isNotMatchAllTemplate) {
-                    if (template.indexPatterns().stream().anyMatch(patternMatchPredicate)) {
+                // For hidden indices, we don't want to return any templates that match "all".
+                boolean match = false;
+                for (String indexPattern : template.indexPatterns()) {
+                    if (Regex.isMatchAllPattern(indexPattern)) {
+                        continue outerLoop;
+                    }
+                    // If we've already found a match, we still want to keep looping to make sure this template doesn't have a "*" pattern.
+                    if (match == false && Regex.simpleMatch(indexPattern, resolvedIndexName)) {
                         candidates.add(Tuple.tuple(name, template));
+                        match = true;
                     }
                 }
+                if (match && exitOnFirstMatch) {
+                    return candidates;
+                }
```

Comment:
One way I think might simplify this function would be to move the  `if (isHidden == false)` block outside the outer loop, given that it's constant within the function. So there would be two `for (Map.Entry<String, ComposableIndexTemplate> entry : templateEntries)` loops, one for hidden and one for not-hidden. The code would be a bit longer, but I think more readable.

More generally, I wonder if the speedup in this function is worth the added complexity with the `match` variable and `continue outerLoop` logic. It seems that the benefit of this is to avoid iterating through indexPatterns twice. Could you keep this more similar to the existing function body, but update it to use `exitOnFirstMatch`? If the speed is more or less the same, it might be worth it to keep complexity low.

Reply:
Yeah I refactored the second inner loop to avoid iterating over `indexPatterns` twice _and_ to avoid using the Stream API in the hot loop. I don't really like the idea of looping over the index patterns twice, but at the same time, I wouldn't be surprised if the majority of templates out there only have a single index pattern.

Reply:
This is no longer relevant after my loop refactoring in [293ff9d](https://github.com/elastic/elasticsearch/pull/106953/commits/293ff9d7ecd187523221204c6f3c50920d51d011).